### DISCUSSION TITLE
Is there a way to control `errorType` in an error response?

### DISCUSSION THREAD

**Discussion by kikuomax** [aws/aws-lambda-rust-runtime#824] [Q&A] [answered]
I would like to handle specific errors from a Lambda handler with a [Retry field in a StepFunctions state machine](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-error-handling.html#error-handling-retrying-after-an-error). To capture errors, I have to specify exact names (`errorType`s) to `ErrorEquals` parameter.

As far as I read the code, `lambda_runtime` uses ~~[`std::any::type_name_of_val`](https://doc.rust-lang.org/std/any/fn.type_name_of_val.html)~~ [`std::any::type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) to produce `errorType` in error responses, which we should avoid using to affect program behavior.
https://github.com/awslabs/aws-lambda-rust-runtime/blob/a4aa23ecd1d98fdba8f5a671f0959ad743bfe138/lambda-runtime/src/lib.rs#L252-L261

Is there any more reliable way to specify `errorType` in error responses?

**Comment by calavera** [comment, marked as answer]
We could introduce some kind of error casting. We could check for it, using `type_name_of_val` as a fallback. That's just an idea, but if you have any other idea, or want to try to implement it, feel free to open a PR.

**Reply by kikuomax** [reply]
@calavera Thanks for answering the question. I will give it a shot!

**Reply by kikuomax** [reply]
- #828

**Comment by evbo** [comment]
@kikuomax In your example how do you define `main`?:
https://github.com/awslabs/aws-lambda-rust-runtime/pull/828/files#diff-ea9f9e8a3273dae249b71b478e9026be91d1ee33c280edec3d1aa25077f7bd11R28

And did you use `Cow` types for any particular reason? There's no real performance penalty in this case, whereas you might want to use `Cow` for a function that may or may not returned a borrow of the original input, presuming the input argument is big and expensive to clone.

**Reply by evbo** [reply]
also, what about `Error` and `Diagnostic` being behind a `"types"` feaure flag?

That way some other crate can be responsible for transforming custom error types into `Diagnostic` that can be reused by multiple lambdas (without a huge compile footprint)

**Reply by kikuomax** [reply]
@evbo Custom error response does not matter to `main` definition. It may be as simple as:
```rust
#[tokio::main]
async fn main() -> Result<(), lambda_runtime::Error> {
    lambda_runtime::tracing::init_default_subscriber();
    let func = lambda_runtime::service_fn(my_handler);
    lambda_runtime::run(function_handler).await?;
    Ok(())
}
```

**Reply by kikuomax** [reply]
I should have included it in the example.
